Model Location 1C Connection Creation

katieperry4 opened the pull request:
<!--- If this is a non-code change, this template is not required; reference any issues or top-level descriptions as needed -->
<!--- All code change PRs should relate to an issue, reference it here; see example below -->
<!--- https://issues.redhat.com/browse/RHOAIENG-123456 -->
Closes:
[RHOAIENG-32187](https://issues.redhat.com/browse/RHOAIENG-32187)
[RHOAIENG-34091](https://issues.redhat.com/browse/RHOAIENG-34091)
[RHOAIENG-31941](https://issues.redhat.com/browse/RHOAIENG-31941)

## Description
<!--- Describe your changes in detail; the what, the why, any findings, etc -->
<!--- Include any screenshots of changed UI; Include any gifs if it was a flow / UX change -->
- Adds connection protocol annotation on secrets created during connection creation and wizard deployment
- Adds checkbox and nameDesc for the connection
- Lets users create a connection in the wizard
- creates connections on deploy

Checkbox doesn't show on existing connections or PVCs:
<img width="932" height="435" alt="Screenshot 2025-10-01 at 4 49 43 PM" src="https://github.com/user-attachments/assets/f3d6df59-356d-45c5-bd05-9b71a636a656" />

Does show on new connections:
<img width="943" height="534" alt="Screenshot 2025-10-01 at 4 50 00 PM" src="https://github.com/user-attachments/assets/e9e5213e-bb94-4552-a103-7dd34bb6cd40" />

Popover:
<img width="530" height="416" alt="Screenshot 2025-10-01 at 4 50 08 PM" src="https://github.com/user-attachments/assets/5f40dd0a-356a-46a0-a5e4-601c28a24b64" />

Unchecked:
<img width="563" height="309" alt="Screenshot 2025-10-01 at 4 50 17 PM" src="https://github.com/user-attachments/assets/9320f787-45c9-438a-a898-7c6ea9180857" />


## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
tested locally

To test:
- Deploy from existing connections (s3, oci, uri) make sure the connections annotation is added to the inference service as well as any additional info like S3 path etc

---------------------------
- Deploy from new connections and create a connection
- enter connection info
- enter name/desc
- deploy and see that a new connection is created with the name you chose and is added to the inference service connection annotation and the info from the connection is on the inference service
- check the created connection has the correct `opendatahub.io/connection-type-protocol` annotation on it

---------------------------
- Do the same but uncheck the create connection box
- see the connection annotation on the inference service and relevant connection info added
- check the created connection (will have a generated name) for the dashboard label to be `false` so it doesn't show in the dashboard and that is has an owner reference to the inference service
- and check the protocol annotation
---------------------------
- Create a connection from workbenches or from the connections tab
- make sure the created connection has the correct `opendatahub.io/connection-type-protocol` annotation on it

## Test Impact
<!--- What tests have you done to cover the implemented functionality -->
<!--- If tests are not applicable, explain why here -->
Added and edited tests

## Request review criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)
- [x] The code follows our [Best Practices](/docs/best-practices.md) (React coding standards, PatternFly usage, performance considerations)

If you have UI changes: 
<!--- You can ignore these if you are doing manifest, backend, internal logic, etc changes; aka non-UI / visual changes -->
- [x] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change.

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - "Create a connection to this location" UI for saving model location connections; deploy flow now does a dry-run, creates real secrets, and uses resolved secret names.
  - Support for Existing/Generated secret flows and selecting URI-based secrets in the wizard.

- **Improvements**
  - Connection protocol annotations applied (uri/s3/oci); smarter prefill and stability in model location flows.
  - Optional test-id for typeahead toggles; refined save-connection checkbox behavior.

- **Bug Fixes**
  - More robust handling when pod status conditions are missing.

- **Tests**
  - Expanded E2E and unit tests covering connection creation, generated-secret flows, validation, and deployment.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

katieperry4 commented on packages/kserve/src/deploy.ts:
```diff
@@ -102,6 +141,9 @@ export const deployKServeDeployment = async (
   }
 
+  // Add owner reference to the secret if the connection is not saved
+  handleSecretOwnerReferencePatch(inferenceServiceData, inferenceService, secretName, dryRun);
```
I think I'm understanding your comment? But I might not be
Right now it's dry running and creating the secret, then dry running and creating the inference service.

Because we can't dry run or create the inference service without the name of a secret that exists so that needs to be done first. So we need the secret name to exist first before we can dry run or create the inference service.

emilys314 commented on frontend/src/api/k8s/secrets.ts:
```diff
@@ -202,3 +203,29 @@ export const deleteSecret = (
+
+export const patchSecretWithInferenceServiceOwnerReference = (
+  secret: SecretKind,
+  inferenceService: InferenceServiceKind,
```
Future extensibility suggestion: I think we can make the owner ref type more generic, because there is nothing model serving related to the concept of k8s owner references
```suggestion
  ownerReference: K8sResourceCommon & { metadata.name: string },
```
or

```suggestion
  inferenceService: K8sResourceIdentifier & {metadata.name: string },
```

(The `& {metadata.name: string }` is required because those two have `name` as an optional property)

emilys314 commented on frontend/src/concepts/modelServingKServe/kserveStatusUtils.ts:
```diff
@@ -84,7 +84,8 @@ export const getInferenceServiceStatusMessage = (
  */
 export const checkModelPodStatus = (model: PodKind): ModelStatus => {
-  const unschedulableCondition = model.status?.conditions.find(
+  const conditions = model.status?.conditions ?? [];
+  const unschedulableCondition = conditions.find(
```
Oh that means our types don't line up with the real world stuff. I'd suggest fixing the type of `PodKind` as well to make it optional `conditions: K8sCondition[];` -> `conditions?: K8sCondition[]; `

emilys314 commented on frontend/src/pages/projects/utils.ts:
```diff
@@ -25,3 +30,12 @@ export const getCustomNotebookSize = (
 export const getPvcAccessMode = (pvc: PersistentVolumeClaimKind): AccessMode =>
   pvc.spec.accessModes[0];
+
+export const getConnectionProtocolType = (
```
thought: 
You could move this into the `src/concepts/connectionTypes/utils.ts` file as I think it's fairly relevant to connection types as a whole.

There also is a `modelServingCompatibleTypesMetadata` obj which does have related metadata info

emilys314 commented on packages/kserve/src/deploy.ts:
```diff
@@ -36,6 +54,7 @@ export type CreatingInferenceServiceObject = {
   runtimeArgs?: RuntimeArgsFieldData;
   environmentVariables?: EnvironmentVariablesFieldData;
   AiAssetData?: AvailableAiAssetsFieldsData;
+  inferenceServiceUID?: string;
```
I don't think i see this being used anywhere?

emilys314 commented on packages/kserve/src/deploy.ts:
```diff
@@ -239,5 +293,108 @@ const applyAnnotations = (
+
+  const connectionTypeName = modelLocationData.connectionTypeObject?.metadata.name ?? 'uri';
+
+  const newConnection = assembleConnectionSecret(
```
I wonder if there is a way to have the connection creation part of the model-serving package?

emilys314 commented on packages/kserve/src/utils.ts:
```diff
@@ -1,4 +1,65 @@
+  inferenceServiceData: CreatingInferenceServiceObject,
+  secretName: string,
+): CreatingInferenceServiceObject => {
+  const newInferenceServiceData = { ...inferenceServiceData };
```
Potential issue: modifying an object's property by reference can lead to problems, especially when dry running is involed

I'd strongly suggest using the `structuredClone()` function

emilys314 commented on packages/kserve/src/utils.ts:
```diff
@@ -1,4 +1,65 @@
+  return newInferenceServiceData;
+};
+
+export const handleSecretOwnerReferencePatch = (
+  inferenceServiceData: CreatingInferenceServiceObject,
+  inferenceService: InferenceServiceKind,
+  secretName: string,
+  dryRun?: boolean,
+): void => {
```
a styling suggestion which helps readability:

I would suggestion still making this an `async (...): Promise<void> =>` for readability. You know that this function is doing some sort of network call since it's a promise. You just don't have to use the `await` when invoking it

That would also help remove in the body of this function the 
```
(async () => {
   ..
)()
```
which i think looks funky

emilys314 commented on packages/model-serving/src/components/deploymentWizard/fields/CreateConnectionInputFields.tsx:
```diff
@@ -0,0 +1,175 @@
+                data-testid="save-connection-checkbox"
+                label="Create a connection to this location"
+                isChecked={createConnectionData.saveConnection}
+                onChange={(_ev, checked) => {
+                  setCreateConnectionData({
+                    ...createConnectionData,
+                    nameDesc: undefined,
+                    saveConnection: checked,
+                  });
+                  resetNameDesc(checked);
+                }}
+              />
```
+1

katieperry4 commented:
@emilys314 I moved the connection creation stuff to just `model-seving` as an extension point and just had the platforms import the functions from `model-serving` in their implementations of the extension
definitely not sure I did it right but I tested it with a new uri and it added the created the secret, added the connection annotation, and added the ownerRef to the secret
I'll do more extensive testing in the morning 😅 and probably some cleanup

emilys314 commented on packages/model-serving/src/components/deploymentWizard/ModelDeploymentWizard.tsx:
```diff
@@ -47,12 +52,41 @@ const ModelDeploymentWizard: React.FC<ModelDeploymentWizardProps> = ({
+    wizardState.state.modelLocationData.data?.connection ||
+    getGeneratedSecretName();
+
+  if (
+    !wizardState.state.createConnectionData.data.nameDesc?.name ||
+    !wizardState.state.createConnectionData.data.nameDesc.k8sName.value
+  ) {
+    wizardState.state.createConnectionData.setData({
+      ...wizardState.state.createConnectionData.data,
+      nameDesc: {
+        name: secretName,
+        description: wizardState.state.createConnectionData.data.nameDesc?.description || '',
+        k8sName: {
+          value: secretName,
+          state: {
+            immutable: false,
+            invalidCharacters: false,
+            invalidLength: false,
+            maxLength: 0,
+            touched: false,
+          },
+        },
+      },
+    });
+  }
```
So I think this change is regarding this comment? https://github.com/opendatahub-io/odh-dashboard/pull/5079#issuecomment-3366125356

The tricky part of avoid useEffect is you have to change the whole order of operations usually. A setState that runs on render is still a side effect happening on a render no matter how it's presented. (In this case having it run raw on the component is a worse react sin)

The only thing a `useEffect` does different from this code, is that `useEffect` won't run on every single rerender thanks to it's dependency array you pass in as the 2nd argument

Griffin-Sullivan replied:
Couldn't you have a useEffect that updates the createConnectionData when the secretName changes?

katieperry4 commented on frontend/src/__tests__/cypress/cypress/tests/mocked/modelServing/modelServingDeploy.cy.ts:
```diff
@@ -611,8 +633,14 @@ describe('Model Serving Deploy Wizard', () => {
+      .findExistingConnectionSelectOption('Test URI Secret')
+      .should('exist')
+      .click();
+    modelServingWizard.findExistingConnectionValue().should('have.value', 'Test URI Secret');
```
Right now I don't add the connections annotation in the dry run bc the deploy and connection creation dry runs run parallel:
```
  Promise.all([
      connectionMethod?.properties.handleConnectionCreation(
        wizardState.state.createConnectionData.data,
        project.metadata.name,
        wizardState.state.modelLocationData.data,
        secretName,
        true, // true = dryrun
      ),
      deployMethod.properties.deploy(
        wizardState.state,
        project.metadata.name,
        undefined,
        serverResource,
        serverResourceTemplateName,
        true, // true = dryrun
      ),
    ]).then(() => {
```
mostly bc if the connection doesn't exist yet the dry run will fail if it has the annotation
so I have it like this in `deploy.ts`
```
  if (createConnectionData?.nameDesc?.name && !dryRun) {
    updatedAnnotations[MetadataAnnotation.ConnectionName] = createConnectionData.nameDesc.name;
  }
```
I could put in a check for whether the ModelLocationType is Existing, but idk if that complicates things too much

Griffin-Sullivan replied:
oh ok makes sense

Griffin-Sullivan commented:
@katieperry4 I think we'll need to create follow up PR to address the edit scenario when you create with a model location that is not a connection. Unless there's a simple way to add in this PR

Griffin-Sullivan commented:
Ok the create connection and create non-connection locations are all working. They are also preloading correctly on edit :+1:

emilys314 commented on frontend/src/components/TypeaheadSelect.tsx:
```diff
@@ -395,7 +396,7 @@ const TypeaheadSelect: React.FunctionComponent<TypeaheadSelectProps> = ({
       variant="typeahead"
       aria-label="Typeahead menu toggle"
-      data-testid="typeahead-menu-toggle"
+      data-testid={dataTestId ?? 'typeahead-menu-toggle'}
```
`dataTestId` is already being used on the `<Select ... />` component because of the `...props` line in 
```
      <Select
        isOpen={isOpen}
        selected={selected}
        onSelect={handleSelect}
        onOpenChange={(open) => !open && closeMenu()}
        toggle={toggle}
        shouldFocusFirstItemOnOpen={false}
        ref={innerRef}
        {...props}
      >
```

emilys314 replied:
I don't think i conveyed it well. But without modifications of this file, yes, the `dataTestId` will not change the "typeahead-menu-toggle", that's because it's used elsewhere in this component already. 

The only change I really care about is that `dataTestId` is not being used in 2 places at the same time. The other usage is here https://github.com/katieperry4/odh-dashboard/blob/9a2291352a17e65568d4147b2b733f336954d132/frontend/src/components/TypeaheadSelect.tsx#L539C9-L539C19

The `toggleProps` is just a suggestion

katieperry4 commented on packages/kserve/src/deployUtils.ts:
```diff
@@ -264,3 +270,37 @@ export const applyModelFormat = (
+  secretName?: string,
+): InferenceServiceKind => {
+  const result = structuredClone(inferenceService);
+  if (createConnectionData.nameDesc?.name) {
+    result.metadata.annotations = {
+      ...result.metadata.annotations,
+    };
```
its the `createConnectionData.nameDesc?.name` not `createConnectionData.saveConnection` so the annotation gets added if there's a name present in `createConnectionData.nameDesc?.name` regardless of whether the user wants to create a connection or not
